File: leetcode-js/317-shortest-distance-from-all-buildings.js
const dirs = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0]
];

/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestDistance = function(grid) {
    // 1. Build a dist matrix filled of 0
    const dist = Array(grid.length).fill().map(() => Array(grid[0].length).fill(0));

    let start = 1;
    let min;

    for (let i = 0; i < grid.length; i += 1) {
        for (let j = 0; j < grid[0].length; j += 1) {
            if (grid[i][j] === 1) {

                // 2. `start` is a way to mark a cell has been visited in this
                // iteration
                start -= 1;

                min = bfsVisit(grid, dist, i, j, start);
            }
        }
    }

    return min === Infinity ? -1 : min;
};

function bfsVisit(grid, dist, row, col, start) {
    const queue = [];

    queue.push([row, col]);

    let level = 0;
    let min = Infinity;

    while (queue.length) {
        // 3. Save current size because we will push elements
        const size = queue.length;

        level += 1;

        for (let k = 0; k < size; k += 1) {
            const node = queue.shift();

            // 4. Move at four directions
            for (const dir of dirs) {
                let newRow = node[0] + dir[0];
                let newCol = node[1] + dir[1];

                // 5. If within boundary, and cell hasn't been visited in this
                // iteration
                if (newRow >= 0 && newRow < grid.length &&
                    newCol >= 0 && newCol < grid[0].length &&
                    grid[newRow][newCol] === start) {

                    queue.push([newRow, newCol]);

                    // 6. Dist contains the sum distance to buildings visited
                    // so far
                    dist[newRow][newCol] += level;

                    min = Math.min(min, dist[newRow][newCol]);
                    grid[newRow][newCol] -= 1;
                }
            }
        }
    }

    return min;
}

console.log(shortestDistance([[1,0,2,0,1],[0,0,0,0,0],[0,0,1,0,0]]))

File: leetcode-java/317-shortest-distance-from-all-buildings.java
public class Solution {
    public int shortestDistance(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] dist = new int[m][n];
        int mark = 0;
        int min = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    min = bfs(grid, m, n, dist, i, j, mark);
                    mark -= 1;
                }
            }
        }

        return min != Integer.MAX_VALUE ? min : -1;
    }

    static int[][] directions = {
        {1, 0},
        {-1, 0},
        {0, 1},
        {0, -1}
    };

    int bfs(int[][] grid, int m, int n, int[][] dist, int i, int j, int mark) {
        Queue<int[]> queue = new LinkedList<>();

        int[] initial = {i, j};
        queue.offer(initial);

        int level = 0;
        int min = Integer.MAX_VALUE;

        while (!queue.isEmpty()) {
            level += 1;
            int size = queue.size();

            for (int k = 0; k < size; k++) {
                int[] pair = queue.poll();

                for (int[] direction : directions) {
                    int x = pair[0] + direction[0];
                    int y = pair[1] + direction[1];

                    if (x >= 0 && x < m && y >= 0 && y < n && grid[x][y] == mark) {
                        dist[x][y] += level;
                        grid[x][y] = mark - 1;
                        min = Math.min(min, dist[x][y]);
                        int[] newPair = {x, y};
                        queue.offer(newPair);
                    }
                }
            }
        }

        return min;
    }
}
